[ML] Skip remote clusters when performing up front privileges validation

przemekwitek opened the pull request:
This PR is a workaround for the bug (inability to check remote indices' privileges via `_has_privileges` API, see https://github.com/elastic/elasticsearch/issues/67798 for more details).
It skips all the remote indices when doing `_has_privileges` check on datafeed and DFA PUT.

Closes https://github.com/elastic/elasticsearch/issues/87832

Commit:
[ML] Skip remote clusters when performing up front privileges validation

Commit:
[ML] Skip remote clusters when performing up front privileges validation

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

elasticsearchmachine commented:
Hi @przemekwitek, I've created a changelog YAML for you.

droberts195 commented on docs/changelog/91813.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 91813
+summary: Skip remote clusters when performing up front privileges validation
+area: Transform
```
I changed the release note to specifically say datafeeds because DFA doesn't currently support CCS.

```suggestion
summary: Skip remote clusters when performing up front privileges validation for datafeeds
area: Machine Learning
```

przemekwitek replied:
I'll apply this suggestion once the release note file re-appears.

przemekwitek replied:
Done.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedManager.java:
```diff
@@ -106,7 +107,11 @@ public void putDatafeed(
         if (XPackSettings.SECURITY_ENABLED.get(settings)) {
             useSecondaryAuthIfAvailable(securityContext, () -> {
-                final String[] indices = request.getDatafeed().getIndices().toArray(new String[0]);
+                final String[] indices = request.getDatafeed()
```
```suggestion
                // TODO: Remove this filter once https://github.com/elastic/elasticsearch/issues/67798 is fixed.
                final String[] indices = request.getDatafeed()
```

przemekwitek replied:
Done.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutDataFrameAnalyticsAction.java:
```diff
@@ -160,8 +163,11 @@ private void putValidatedConfig(
         if (securityContext != null) {
             useSecondaryAuthIfAvailable(securityContext, () -> {
                 final String username = securityContext.getUser().principal();
+                final String[] sourceIndices = Arrays.stream(preparedForPutConfig.getSource().getIndex())
```
```suggestion
                // DFA doesn't support CCS, but if it did it would need this filter, so it's safest to have the filter
                // in place even though it's a no-op.
                // TODO: Remove this filter once https://github.com/elastic/elasticsearch/issues/67798 is fixed.
                final String[] sourceIndices = Arrays.stream(preparedForPutConfig.getSource().getIndex())
```

przemekwitek replied:
Done.

elasticsearchmachine commented:
Hi @przemekwitek, I've updated the changelog YAML for you.

droberts195 reviewed (commented):
Would it be possible to add a test for the datafeed case?

I don't think it will be possible to add a test for the DFA case as CCS is banned elsewhere in the validation. But I'm happy to leave the DFA code in this PR so this problem doesn't reappear if DFA is ever changed to support CCS.

Commit:
Apply review comments

Commit:
Apply review comments

przemekwitek commented:
> Would it be possible to add a test for the datafeed case?

Yes, I think so. 

> I don't think it will be possible to add a test for the DFA case as CCS is banned elsewhere in the validation. But I'm happy to leave the DFA code in this PR so this problem doesn't reappear if DFA is ever changed to support CCS.

+1

juliaElastic committed:
[Fleet] Added logs-elastic_agent* read privileges to kibana_system (#91701)

* Added logs-elastic_agent* read privileges to kibana_system

* Update docs/changelog/91701.yaml

* added unit test

* Fixed formatting

* removed read cross cluster role

Commit:
Apply review comments

Commit:
Apply review comments

original-brownbear committed:
Build more compact RepositoryData when parsing from JSON (#91817)

Low effort imporovement to #89952 mostly.

We should be turning the index identifier lookup into an immutable map
when parsing these right away. We do this conversion/copy for both the adding
and removing of snapshots from the `IndexMetadataGenerations` later on anyways
so this doesn't add any CPU cost overall. What it does however is save a massive
amount of heap for single index snapshots (where the overhead of hash map over
the immutable map is the greatest) when first parsing this structure from the repo
and potentially having it duplicated on heap many times over due to #89952.

philippkahr committed:
Metricbeat monitoring collection suggest to enable local exporters (#77409)

* Metricbeat monitoring collection suggest to enable local exporters

Those steps are not needed.

1. We enable the local _legacy_ exporters
2. In the last step, we disable the local _legacy_ exporters again.

* Also remove legacy collection step from Filebeat page

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

albertzaharovits committed:
Consolidate handling of "all" for index expression resolver (#91735)

A refactoring such that callers of
IndexNameExpressionResolver#resolveExpressions don't
have to massage parameters when resolving _all and
its ilk (empty, null, *).

elasticsearchmachine commented:
Hi @przemekwitek, I've created a changelog YAML for you.

Commit:
Update docs/changelog/91813.yaml

Commit:
Update docs/changelog/91813.yaml

Commit:
Apply review comment

Commit:
Apply review comment

martijnvg committed:
Big arrays sliced from netty buffers (long) (#91641)

Based on #90745 but for longs. This should allow aggregations down the
road to read long values directly from netty buffer, rather than copying
it from the netty buffer.

Relates to #89437

przemekwitek commented:
run elasticsearch-ci/packaging-tests-windows-sample

pgomulka committed:
Improve InnocuousThread permission checks handling (#91704)

on shutdown, the jdk's InnocuousThread can try to change a thread name. This requires "java.lang.RuntimePermission" "modifyThread" permission. However InnocuousThread doe not inherit any Access Control Context and therefore have no permissions. This results in AccessControlException.
This commit fixes this by skipping a check for modify thread permission if a thread is innocuous.
relates #91658 and #91650

When previously described AccessControlException is thrown, it is not being catched anywhere in the Elasticsearch, hence it ends up being handled by ElasticsearchUncaughtExceptionHandler#onNonFatalUncaught
This is being again being run by the thread [process reaper] which is an innocuous thread (jdk specific) and has no permissions. onNonFatalUncaught is trying to log a message, but this in turn requires java.lang.RuntimePermission" "getenv.*" permission. which is does not have. This again results in AccessControlException java.lang.RuntimePermission" "getenv.*"

We can fix this by executing with doPrivileged in ElasticsearchUncaughtExceptionHandler#onNonFatalUncaught and this will stop the Security Manager's walk and will have ES's global grant permissions.

closes #91650

davidkyle committed:
[ML] Rename semantic search input parameter (#91787)

Formerly `query_string` now `model_text`

fcofdez committed:
Fix DataStreamIT#testWriteLoadAndAvgShardSizeIsStoredInABestEffort (#91655)

Closes #91384

pxsalehi committed:
Wait for task on master in testGetMappingsCancellation (#91709)

dimitris-athanasiou committed:
[ML] Increase timeout for _infer in PyTorchModelIT (#91853)

Inference requests running on slow workers in CI may exceed the
default 10s timeout. This commit increases the timeout so those
tests become more robust.

original-brownbear committed:
Make wiping all snapshots in REST tests more efficient (#91828)

The logic here was a left-over from a time when we couldn't bulk delete snapshots.
We can simply run a wildcard delete per fs-repo now and it will transparently take
care of in-progress snapshots etc. No need for any of the logic retrying on existing
in-progress snapshots and such.

original-brownbear committed:
Fix CacheServiceTests.testProcessShardEviction (#91855)

Slightly adjust the mechanics of the map of futures so that `waitForCacheFilesEvictionIfNeeded` and
a future obtained from `pendingShardsEvictions` will always be
consistent since both the future and the map removal happen under the
same mutex. The existing approach of using the future from the executor
submit and removing it from the map during the execution of its task
would lead to a race where the future is always removed from the map
before it completes, so `waitForCacheFilesEvictionIfNeeded` would
briefly not see the future in the map and return while the future has
not yet been completed.

closes #91814

thecoop committed:
Add trace.id to request trace logs (#91772)

This adds trace.id, extracted from traceparent header, so its present in the request trace log. It is already included in response trace logs via the threadcontext, but that is not set at request log time.

andreidan committed:
[DOCS] disable the ILM history store on full cluster restore (#88515)

original-brownbear committed:
Simplify and optimize deduplication of RepositoryData for a non-caching repository instance (#91851)

This makes use of the new deduplicator infrastructure to move to more
efficient deduplication mechanics.
The existing solution hardly ever deduplicated because it would only
deduplicate after the repository entered a consistent state. The
adjusted solution is much simpler, in that it simply deduplicates such
that only a single loading of `RepositoryData` will ever happen at a
time, fixing memory issues from massively concurrent loading of the repo
data as described in #89952.

closes #89952

williamrandolph committed:
Check stable plugin version at install and load time (#91780)

For a node to install a stable plugin, the node's Elasticsearch version must have the same major version as the Elasticsearch libraries against which the stable plugin was built. Additionally, a node will reject stable plugins that have been built with a future version of Elasticsearch.

* Add version checks for stable plugins
* Update docs/changelog/91780.yaml

slobodanadamovic committed:
Add ability to resolve the cluster alias of remote cluster connections (#91724)

This commit adds the ability to resolve a cluster alias 
of a `Transport.Connection` to `RemoteConnectionManager` class.

Commit:
Merge branch 'fix_has_privileges_df_bug' of https://github.com/przemekwitek/elasticsearch into fix_has_privileges_df_bug

Commit:
Fix creating index privileges

Commit:
Add multi cluster tests with security for ML

Commit:
Adapt multi cluster test to ML

Commit:
Rename test files
